ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryRegistryImpl.java

```diff
@@ -0,0 +1,366 @@
+        unregisterQuery(oldQuery);
+      }
+
+      persistentQueries.put(queryId, persistentQuery);
+      if (createAsQuery) {
+        createAsQueries.put(persistentQuery.getSinkName(), queryId);
+      } else {
+        // Only INSERT queries exist beside CREATE_AS
+        sinkAndSources(persistentQuery).forEach(sourceName ->
+            insertQueries.computeIfAbsent(sourceName,
+                x -> Collections.synchronizedSet(new HashSet<>())).add(queryId));
+      }
+    }
+    allLiveQueries.add(query);
+    notifyCreate(serviceContext, metaStore, query);
+    query.initialize();
+  }
+
+  private void unregisterQuery(final QueryMetadata query) {
+    if (query instanceof PersistentQueryMetadata) {
+      final PersistentQueryMetadata persistentQuery = (PersistentQueryMetadata) query;
+      final QueryId queryId = persistentQuery.getQueryId();
+      persistentQueries.remove(queryId);
+
+      // If query is a INSERT query, then this line should not cause any effect
+      createAsQueries.remove(persistentQuery.getSinkName());
+
+      // If query is a C*AS query, then these lines should not cause any effect
+      sinkAndSources(persistentQuery).forEach(sourceName ->
+          insertQueries.computeIfPresent(sourceName, (s, queries) -> {
+            queries.remove(queryId);
+            return (queries.isEmpty()) ? null : queries;
+          })
+      );
+    }
+
+    allLiveQueries.remove(query);
+    notifyDeregister(query);
+  }
+
+  private void notifyCreate(
+      final ServiceContext serviceContext,
+      final MetaStore metaStore,
+      final QueryMetadata queryMetadata) {
+    this.eventListeners.forEach(l -> l.onCreate(serviceContext, metaStore, queryMetadata));
+  }
+
+  private void notifyStateChange(
+      final QueryMetadata queryMetadata,
+      final KafkaStreams.State before,
+      final KafkaStreams.State after
+  ) {
+    this.eventListeners.forEach(l -> l.onStateChange(queryMetadata, before, after));
+  }
+
+  private void notifyError(final QueryMetadata queryMetadata, final QueryError error) {
+    this.eventListeners.forEach(l -> l.onError(queryMetadata, error));
+  }
+
+  private void notifyClose(final QueryMetadata queryMetadata) {
+    this.eventListeners.forEach(l -> l.onClose(queryMetadata));
+  }
+
+  private void notifyDeregister(final QueryMetadata queryMetadata) {
+    this.eventListeners.forEach(l -> l.onDeregister(queryMetadata));
+  }
+
+  private Iterable<SourceName> sinkAndSources(final PersistentQueryMetadata query) {
+    return Iterables.concat(
+        Collections.singleton(query.getSinkName()),
+        query.getSourceNames()
+    );
+  }
+
+  @FunctionalInterface
+  interface QueryExecutorFactory {
+    QueryExecutor create(
+        SessionConfig config,
+        ProcessingLogContext processingLogContext,
+        ServiceContext serviceContext,
+        FunctionRegistry functionRegistry
+    );
+  }
+
+  private class PerQueryListener implements QueryMetadata.Listener {
```

Comment:
We have two `PerQueryListener` now, one is here and another one is inside `QueryStateListener`, and the call trace is `QueryMetadata.PerQueryListener -> QueryMetadata#notifyXXX -> QueryStateListener.PerQueryListener`. 

Could we rename them differently, e.g. rename this one as just `ListenerImpl` since this is not supposed to be customized and only exposed as an interface for testing purposes.